### PR TITLE
improve Drop documentation

### LINKED ISSUE #36073: std::ops::Drop documentation focuses on scope, ignores other causes of drops
There was some confusion in #rust over when values are dropped. It was pointed out that the `std::ops::Drop` documentation only says drop is called if "a value goes out of scope", which is inaccurate:
- Scopes apply to variables, not values
- Values can be dropped for other reasons than scopes ending. For example, if a variable is re-assigned, the old value is dropped. There are also temporaries.

I don't know how to best phrase it, but focusing only on scopes is evidently a good way to confuse people.

### PR DESCRIPTION
**Pull request by lcnr**
Fixes #36073

This is a continuation of #57449 and most of the work here was done by
the excellent @steveklabnik.

### CODE REVIEW

File: src/libcore/ops/drop.rs
Comment (lcnr) [inline src/libcore/ops/drop.rs:120]: IMO we can/should probably replace this with.

> The order in which values are dropped is always specified and can be viewed in **the reference**.

I believe that I did not care about the drop order as I started learning rust. So for me this would probably be the least relevant section.

File: src/libcore/ops/drop.rs
Comment (crlf0710) [inline src/libcore/ops/drop.rs:1]: I think this is still a little inaccurate, `Drop` itself it not the destructor, but a "hook" for destructor. Maybe change it to something like "Custom code within the destructor for cleaning up resources."

File: src/libcore/ops/drop.rs
Comment (crlf0710) [inline src/libcore/ops/drop.rs:11]: maybe "to customize the destructor of a type."

File: src/libcore/ops/drop.rs
Comment (crlf0710) [inline src/libcore/ops/drop.rs:21]: Maybe refer to something like `drop_in_place` instead of `Drop::drop` directly here. By implementing `Drop::drop` the behavior of `drop_in_place` is customized. cc. @RalfJung

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs:1]: Changed it to `Custom code within the destructor`. `Drop` is also used for panic guards (which don't feel like a resource to me)

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs:21]: hm, I don't want to also use `drop_in_place`, even if it is slightly more correct. :thinking: 

For new users the difference seems to be negligible and causes them to look at one more place. We already have `Drop::drop` and `mem::drop` here, which is more complicated than most people expect when first learning about `drop`.

Comment (RalfJung) [reply src/libcore/ops/drop.rs:21]: I think the text here is incorrect though, and misleading. If anything it should say that a call to `mem::drop` is inserted.

For example, when you have your own type like
```rust
struct Foo(Vec<u8>);
```
and you `impl Drop for Foo`, then calling `Drop::drop` would *not* drop the vector! `Drop::drop` is shallow, `mem::drop`/`drop_in_place` are deep. I do think this is something that must be explained.

IMO it makes more sense to talk about `Drop::drop` and `drop_in_place`, and leave out `mem::drop` (it can be explained as an afterthought, as a simple safe wrapper around `drop_in_place`). `Drop::drop` drops "one layer" of data structures, and `drop_in_place` adds a bunch of auto-generated code to recursively drop each layer of nested structs/enums.

Comment (lcnr) [reply src/libcore/ops/drop.rs:21]: The more I think about this the less I wan to use `drop_in_place` here. This method is unsafe and its documentation is not really beginner friendly (which is ok, as no beginner should have to use it tbh).

> and you impl Drop for Foo, then calling Drop::drop would not drop the vector! Drop::drop is shallow, mem::drop/drop_in_place are deep. I do think this is something that must be explained.

This is explained in the section `Dropping is done recursively` further down. I would like to move this section upwards, but the section `You cannot call Drop::drop yourself` is also important, so I don't want to move that one further down. :sweat_smile:

Comment (RalfJung) [reply src/libcore/ops/drop.rs:21]: > This method is unsafe and its documentation is not really beginner friendly (which is ok, as no beginner should have to use it tbh).

Maybe its doc should be improved? IMO properly understanding drop requires understanding `drop_in_place`. That might not be beginner-friendly, but that is because dropping is actually non-trivial.

Saying "the compiler automatically insert a call to `Drop::drop`" is factually incorrect and thus I think we have to say something else. I appreciate sometimes it's helpful to say something "almost correct" to ease understanding (I do that all the time), but what it currently says is too wrong to be useful, IMO. Going into `drop_in_place` is just one way to fix that (it's what I'd do, but I am open to alternatives).

Comment (lcnr) [reply src/libcore/ops/drop.rs:21]: My current ideas which use  `mem::drop` instead of `Drop::drop` are either more incorrect or more complex. :thinking:

Comment (RalfJung) [reply src/libcore/ops/drop.rs:21]: IMO saying "it automatically inserts calls to `mem::drop`" would be less incorrect than what it says now. `mem::drop` is "move followed by `drop_in_place`"; this would "just" be "lying" about the move part -- in contrast to the current text which entirely conflates user-written drop impls with auto-generated drop glue.

Comment (lcnr) [reply src/libcore/ops/drop.rs:21]: > I appreciate sometimes it's helpful to say something "almost correct" to ease understanding (I do that all the time when writing papers), but what it currently says is too wrong to be useful, IMO.

hmm, how about

---


When a value goes out of scope, the compiler automatically insert a call to `Drop::drop` for that value, and then recursively for all other values it contains:

```rust
struct HasDrop;

impl Drop for HasDrop {
    fn drop(&mut self) {
        println!("Dropping HasDrop!");
    }
}

struct HasTwoDrops {
    one: HasDrop,
    two: HasDrop,
}

impl Drop for HasTwoDrops {
    fn drop(&mut self) {
        println!("Dropping HasTwoDrops!");
    }
}

fn main() {
    let _x = HasTwoDrops { one: HasDrop, two: HasDrop };
}
```

Running this will print

```text
Dropping HasTwoDrops!
Dropping HasDrop!
Dropping HasDrop!
```

Even if we remove the implementation of `Drop` for `HasTwoDrop`, the destructors of its fields
are still called. This would result in

```test
Dropping HasDrop!
Dropping HasDrop!
```

Comment (RalfJung) [reply src/libcore/ops/drop.rs:21]: > When a value goes out of scope, the compiler automatically insert a call to Drop::drop for that value, and then recursively for all other values it contains

That is at least not incorrect any more. :D 

But I still think the docs should *name* that concept of "call Drop::drop for that value, and then recursively for all other values it contains". Naming things makes explaining and remembering them much easier. The name I have used  recentlyfor that concept is `drop_in_place`. Sometimes it's also called "drop glue", though I am not sure if that is strictly the same thing (maybe "drop glue" is just the code that does the gluing, as opposed to the entire operation). I am open for other suggestions for what this should be called, but I feel rather strongly that it should be called *something*.

Comment (lcnr) [reply src/libcore/ops/drop.rs]: I now use "drop glue" there, updated the PR. 

> But I still think the docs should name that concept of "call Drop::drop for that value, and then recursively for all other values it contains".

I think that I personally use "dropping" for this, i.e. when "dropping" a value, the compiler calls `Drop::drop` for it and all its owned values.

Comment (RalfJung) [reply src/libcore/ops/drop.rs]: > I now use "drop glue" there, updated the PR.

I like that terminology. :D
I'll leave some specific comments about the structure.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs]: Why did you remove these examples?

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:21]: This is the first paragraph in a section titled

> ## Implementing `Drop`

I don't think that is good. A large part of the confusion (at least for me) around drop is that we have `Drop` impls and "drop glue", and they are not the same thing. So If eel there should be a section titled `## Drop Glue` that explains what drop glue is, separately (and maybe after? not sure) from the section that explains `impl Drop`.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [reply src/libcore/ops/drop.rs]: Ah you have more examples below, maybe they are not needed any more.

Comment (lcnr) [reply src/libcore/ops/drop.rs]: This documentation is already quite long and I think that `This means calling to `Drop::drop` for it, and then recursively for all other values it contains:` + the example with the expected output are hopefully enough

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs:21]: true

Comment (lcnr) [reply src/libcore/ops/drop.rs:21]: I though about this some more and am still not completely satified.

I first considered splitting "implementing" and "using/drop glue" into 2 different sections,
but was unable to think of a way to explain one of them without requiring the other one.

I renamed `Implementing Drop` to `Using the Drop trait` for now, which makes at least some sense :thinking:

Comment (RalfJung) [reply src/libcore/ops/drop.rs:21]: Naively I thought explaining the trait first should work, but probably I am missing a dependency?

### Implementing `Drop`

To customize the operations that are automatically performed when a value goes out of scope, ...

### Drop glue

Whenever a value goes out of scope, the compiler automatically inserts "drop glue" which calls the `Drop` impl of that value (if it exists), and then recursively calls the "drop glue" for each field of a `struct`/`enum`.

Comment (lcnr) [reply src/libcore/ops/drop.rs:21]: > ## Implementing `Drop`
>
>To customize the operations that are automatically performed when a value goes out of scope, ...

My problem with this approach is that without mentioning`DropGlue`, while reading the "Implementing `Drop`" section one can reasonably think that one has to manually call drop for the `Vec` in `struct A(Vec<u8>)`.

So while I could add an example just showing how to implement this trait without explaining how destructors work, I think that this would cause more harm than good.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:13]: "Most notably when dealing with resources" is a sentence without an active verb, that should be fixed.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:22]:
```suggestion
/// for this value. This means calling `Drop::drop` for it, and then recursively for all
```

Comment (RalfJung) [reply src/libcore/ops/drop.rs]: Actually I'd formulate this a bit differently:

> This means it will call `Drop::drop` on that value, and then recursively call the "drop glue" for all fields.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [reply src/libcore/ops/drop.rs:21]: Hm, I see. In that case the order of things you have probably makes sense. But the section here now feels like it mainly explains how the `Drop` trait is called by the compiler, so shouldn't the section title be "Drop glue"?

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:3]: So here we call it "destructor" and later we call the same thing "drop glue"?

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:73]: That argument is... a bit strange to me. I can call `mem::drop` myself without double-dropping, so if that was the problem, why wouldn't the compiler just apply the same rules for `Drop::drop`?

The answer is that *the type of `Drop::drop` is a lie*. That function is *not* like any other function that takes a mutable reference. Usually, a function taking `&mut T` borrows `T` and then it has to give back an intact `T` when the borrow is over. `Drop::drop`, on the other hand, partially *destroys* the `T`, and it may rely on the fact that no other method is called on this type any more after it returns. *That* is why it cannot be called directly.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [reply src/libcore/ops/drop.rs:3]: "destructor" could be another term we could use for "drop glue", but I am not sure if that isn't confusing as people might also think it refers to `Drop::drop`. But in that case we should later explain which of the two concepts that we explain in more detail, is meant by "destructor" -- or probably we should never use that term again, and only use it here in the introduction before we established the terminology.

In fact, that made me think of a structure like that:

> When a value is no longer needed, Rust will run a "destructor" on that value.
>
>  [...]
>
> Destructors in Rust consist of two components:
> * Implementations of the `Drop` trait, that customize the destructor for values of a particular type.
> * The compiler-generated "drop glue", that automatically calls `Drop::drop` on a value and then recursively does the same for all its fields.

So to avoid the cyclic dependency, we basically introduce both terms at the same time. What do you think?

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs]: done

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs:73]: :+1:

File: src/libcore/ops/drop.rs
Comment (RalfJung) [reply src/libcore/ops/drop.rs]: (I posted another comment above which got hidden by GH when you accepted the suggestions... but anyway this is superseded by https://github.com/rust-lang/rust/pull/71598#discussion_r417826595)

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs:3]: yeah :sweat_smile:

I now rewrote the section mentioning "drop glue"

Comment (lcnr) [reply src/libcore/ops/drop.rs:3]: Didn't see your second comment, seems like a good approach :+1:

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs]: improved

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs]: Don't really like "drop glue" as a section title. "Using the `Drop` trait" is not optimal, but I think it works well enough. This section shows both how `Drop` is used by the compiler and how to implement it yourself.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:75]:
```suggestion
/// the method has been called. This is not reflected in the type of
/// `Drop::drop`, which does not take ownership of its input. Thus
/// Rust prevents misuse by not allowing you to call `Drop::drop` directly.
```

Comment (lcnr) [reply src/libcore/ops/drop.rs]: :+1: thanks, will squash this before it gets merged

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:11]: This is the first place where `Drop` comes up -- and I feel the best way to explain what it does is to actually have that text you have below, right here instead:
```
/// Destructors consist of two components:
/// - If the special `Drop` trait is implemented for the type of the value, a call to `Drop::drop`.
/// - The automatically generated "drop glue", that recursively calls the destructors
///     of the all fields of this value.
///
/// In other words, the `Drop` trait lets us add our own code to the destructor of a type. [...]
```
Right now, the text just very vaguely says `Drop` is "a way to customize the destructor", but is it the only way and how is it customized anyway? It's not complicated, so I feel we can just say it upfront instead of leaving the reader wonder.

Comment (lcnr) [reply src/libcore/ops/drop.rs]: :+1: I moved that section upwards  and slightly rewrote the "Using the `Drop` trait" section.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:27]: I am not sure what this slightly ominous announcement is telling me. It doesn't really help to understand the following text, does it?

"Examples" honestly seems like a good section title with the current structure. :D

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:12]:
```suggestion
/// - A call to `Drop::drop` for that value, in case the special `Drop` trait is implemented for its type.
```
This is the first time `Drop` is mentioned, might be a good idea to introduce it as a "special trait"?

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs:27]: :crab: To strike fear into their hearts! Fear the crab :crab:

Comment (RalfJung) [reply src/libcore/ops/drop.rs:27]: :rofl: 
I'd just not mention the restrictions until the subsection dedicated to it -- it's unnecessarily distracting.

Comment (lcnr) [reply src/libcore/ops/drop.rs]: Fixed

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs]: I don't like `special` here. I feel like

> I am not sure what this slightly ominous announcement is telling me. It doesn't really help to understand the following text, does it?

also applies here.

Comment (RalfJung) [reply src/libcore/ops/drop.rs]: Hm. I feel like the text here already says why it is special -- because it is called by the destructor.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [inline src/libcore/ops/drop.rs:26]:
```suggestion
/// To see destructors in action, let's take a look at the following program:
```
There's no "second" here, is there?

Comment (lcnr) [reply src/libcore/ops/drop.rs:26]: I don't want to waste one second here :+1:

File: src/libcore/ops/drop.rs
Comment (lcnr) [reply src/libcore/ops/drop.rs]: Hmm, it feet like your wording may imply that there's some other `Drop` trait which is `special`.
Changed "the" to "this".

> A call to `Drop::drop` for that value, if this special `Drop` trait is implemented for its type.

Comment (RalfJung) [reply src/libcore/ops/drop.rs]: To me "this special drop trait" implies that there's another one ("this X and that X", making it multiple X), but "*the* special drop trait" pretty clearly says that there is only one.

Comment (lcnr) [reply src/libcore/ops/drop.rs]: hmm, there are other `Drop` traits. for me "the" does not express that we are talking about the currently documented item. When reading your recommendation I thought that there may exist another `Drop` trait which is special, while this `Drop` trait is somehow not.
```rust
trait Drop { } // this is valid and not used in destructors
```

**Comment by Mark-Simulacrum** [comment]
@RalfJung -- do you want to be the reviewer for this PR? It looks like you've already invested some time into it. If not I'm happy to try to conclude the review here and approve.

**Comment by RalfJung** [comment]
I mostly helped refine the language and wording here. I can help finish this but would not be comfortable r+'ing alone -- we don't have a test suite for docs stuff like this so I think it is best to get as many eyes on it as we can. :)

**Review by RalfJung — APPROVED** [review]
LGTM (I'd word some things differently but arguably that's just taste).

Would be good to get another set of eyes on it though.

**Comment by Mark-Simulacrum** [comment]
Okay I can try to be that set of eyes :)

File: src/libcore/ops/drop.rs
Comment (Mark-Simulacrum) [inline src/libcore/ops/drop.rs:133]: I would not say "don't manage resources" here -- that's not *necessarily* true, though would be quite unusual. I don't think we need to be specific why this restriction is in place, just mentioning that it exists seems good enough.

**Review by Mark-Simulacrum — COMMENTED** [review]
This looks good to me, though I will note that it is much more conversational than I think the standard library docs are generally. I don't personally mind that too much, though I do think that we may want to stay more consistent. Regardless, this seems like a good improvement so I'm happy to approve; let's squash the commits into one and if you can fix the one nit I left then we can merge!

File: src/libcore/ops/drop.rs
Comment (liigo) [reply src/libcore/ops/drop.rs:85]: Many (if not most) C/C++ people care drop order.

File: src/libcore/ops/drop.rs
Comment (RalfJung) [reply src/libcore/ops/drop.rs:133]: If we do say anything, we could say that `Copy` types get duplicated by the compiler in an uncontrolled way so it would be very hard to predict *how often* the destructor runs, which can easily lead to subtle bugs. To avoid these bugs, we forbid `Copy` types from having a destructor

Comment (lcnr) [reply src/libcore/ops/drop.rs]: I think that a short explanation is helpful here. Changed it to be closes to @RalfJung 's explanation :sparkles:

**Comment by Mark-Simulacrum** [comment]
Thanks!

@bors r=RalfJung,Mark-Simulacrum

**Comment by RalfJung** [comment]
MSYS installation failure (Cc @pietroalbini -- or should I ping the entire infra team in such cases?)

@bors retry

**Comment by pietroalbini** [comment]
@RalfJung ping the whole team. This was fixed in #71995.

**Pull request merged**
